### PR TITLE
New module: Add module to collect facts for Amazon Redshift Cluster (cloud/amazon/redshift_facts)

### PR DESCRIPTION
**Pull request by j-carl**
Gather facts about Redshift Cluster(s).

##### SUMMARY

Gather facts about Redshift Cluster(s) either by:

- Get all cluster(s)
- Get cluster(s) by name (uses a regular expression)
- Get cluster(s) by tags

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

`lib/ansible/modules/cloud/amazon/redshift_facts.py`

##### ANSIBLE VERSION

```
ansible 2.2.1.0
  config file =
  configured module search path = Default w/o overrides
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/redshift_facts.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+                    'metadata_version': '1.0'}
+
+DOCUMENTATION = '''
+---
+module: redshift_facts
+author: "Jens Carl (@j-carl)"
+short_description: Gather facts about Redshift cluster(s)
+description:
+  - Gather facts about Redshift cluster(s)
+version_added: "2.3"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/amazon/redshift_facts.py]: `"2.4"`

File: lib/ansible/modules/cloud/amazon/redshift_facts.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+from ansible.module_utils.ec2 import HAS_BOTO3, camel_dict_to_snake_dict
+
+try:
+    from botocore.exception import ClientError
+except ImportError:
+    pass  # will be picked up from imported HAS_BOTO3
+
+
+def match_tags(tags_to_match, cluster):
+    for key, value in tags_to_match.item():
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift_facts.py]: I think this was a typo- should be tags_to_match.items():

File: lib/ansible/modules/cloud/amazon/redshift_facts.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+def main():
+
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(
+        dict(
+            cluster_identifier=dict(type='str', aliases=['identifier', 'name']),
+            tags=dict(type='dict')
+        )
+    )
+    module = AnsibleModule(argument_spec=argument_spec)
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift_facts.py]: you could add supports_check_mode=True to AnsibleModule

File: lib/ansible/modules/cloud/amazon/redshift_facts.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+
+    for cluster in clusters['Clusters']:
+
+        matched_identifier = True
+        if identifier:
+            matched_identifier = identifier_prog.search(cluster['ClusterIdentifier'])
+
+        matched_tags = True
+        if tags:
+            matched_tags = matched_tags(tags, cluster)
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/redshift_facts.py]: Another typo - should be match_tags(tags, cluster)

**Review by s-hertel — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/redshift_facts.py (hunk 2 above)
Comment (j-carl) [reply lib/ansible/modules/cloud/amazon/redshift_facts.py]: Thx and I fixed it

File: lib/ansible/modules/cloud/amazon/redshift_facts.py (hunk 4 above)
Comment (j-carl) [reply lib/ansible/modules/cloud/amazon/redshift_facts.py]: Thx

File: lib/ansible/modules/cloud/amazon/redshift_facts.py (hunk 3 above)
Comment (j-carl) [reply lib/ansible/modules/cloud/amazon/redshift_facts.py]: I added the feature

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
shipit

**Comment by abadger** [comment]
Merged to devel for 2.4.0 release.

**Pull request merged**